server/lib/realtime/workflows/manager.ex

```diff
@@ -0,0 +1,153 @@
+defmodule Realtime.Workflows.Manager do
+  use GenServer
+  require Logger
+
+  alias Realtime.Adapters.Changes
+  alias Realtime.TransactionFilter
+  alias Realtime.Workflows
+
+  defmodule State do
+    defstruct [:state, :queue, :table]
+  end
+
+  @table_name :workflows_manager
+
+  ## Manager API
+
+  def start_link(config) do
+    GenServer.start_link(__MODULE__, config, name: __MODULE__)
+  end
+
+  @doc """
+  Send notification for txn to the workflow manager.
+  """
+  def notify(txn) do
+    GenServer.cast(__MODULE__, {:notify, txn})
+  end
+
+  @doc """
+  Return a list of workflows that can be triggered by change.
+  """
+  def workflows_for_change(txn) do
+    # No need to call GenServer, we can lookup the table directly
+    :ets.foldl(
+      fn ({_, workflow}, acc) ->
+        event = %{event: "*", relation: workflow.trigger}
+        if TransactionFilter.matches?(event, txn) do
+          [workflow | acc]
+        else
+          acc
+        end
+      end,
+      [],
+      @table_name
+    )
+  end
+
+  @doc """
+  Return the workflow with the given id, or nil if not found.
+  """
+  def workflow_by_id(id) do
+    case :ets.lookup(@table_name, id) do
+      [workflow] -> workflow
+      [] -> nil
+    end
+  end
+
+  ## GenServer Callbacks
+
+  @impl true
+  def init(config) do
+    workflows = :ets.new(@table_name, [:named_table, :protected])
+    state = %State {
+      state: :init,
+      queue: [],
+      table: workflows
+    }
+
+    Task.start_link(fn () ->
+      workflows = Workflows.list_workflows()
+      GenServer.cast(__MODULE__, {:load_workflows, workflows})
+    end)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_cast({:load_workflows, workflows}, %State{state: :init, queue: queue, table: table} = state) do
+    Enum.each(workflows, fn workflow_data ->
+      workflow = %{
+        id: workflow_data.id,
+        trigger: workflow_data.trigger,
+        definition: workflow_data.definition
+      }
+      :ets.insert(table, {workflow.id, workflow})
+    end)
+    Enum.each(queue, fn txn -> do_handle_notification(table, txn) end)
+    new_state = %State{state | state: :sync, queue: []}
+    {:noreply, new_state}
+  end
```

Comment:
@fracek you can simplify this with `init` returning `{:ok, state, {:continue, :load_workflows}}` and `handle_continue`:

```suggestion
  @impl true
  def init(_config) do
    :ets.new(@table_name, [:named_table, :protected]) # returns @table_name
    {:ok, %{}, {:continue, :load_workflows}}
  end

  @impl true
  def handle_continue(:load_workflows, state) do
    Workflows.list_workflows()
    |> Enum.each(fn workflow_data ->
      workflow = %{
        id: workflow_data.id,
        trigger: workflow_data.trigger,
        definition: workflow_data.definition
      }
      :ets.insert(table, {workflow.id, workflow})
    end)

    {:noreply, state}
  end
```

This way, `handle_continue` ensures that all workflows are loaded before other messages are processed and you don't have to keep track of `state` and `queue`. You don't even need `table` in state as you can just reference `@table_name`.

Reply:
Great suggestion, I will add this.

Reply:
Now I wonder if it even needs to be a genserver, we only need to initialise the ets table which can be done once at startup.

Reply:
@fracek I think you're right!